packages/puppeteer-core/src/bidi/core/Browser.ts

```diff
@@ -0,0 +1,162 @@
+/**
+ * @license
+ * Copyright 2024 Google Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import type * as Bidi from 'chromium-bidi/lib/cjs/protocol/protocol.js';
+
+import {EventEmitter} from '../../common/EventEmitter.js';
+import {throwIfDisposed} from '../../util/decorators.js';
+
+import type BrowsingContext from './BrowsingContext.js';
+import type {SharedWorkerRealm} from './Realm.js';
+import type Session from './Session.js';
+import UserContext from './UserContext.js';
+
+/**
+ * @internal
+ */
+export type AddPreloadScriptOptions = Omit<
+  Bidi.Script.AddPreloadScriptParameters,
+  'functionDeclaration' | 'contexts'
+> & {
+  contexts?: [BrowsingContext, ...BrowsingContext[]];
+};
+
+/**
+ * @internal
+ */
+export default class Browser extends EventEmitter<{
+  /** Emitted after the browser closes. */
+  closed: {
+    /** The reason for closing the browser. */
+    reason: string;
+  };
+  /** Emitted after the browser disconnects. */
+  disconnected: {
+    /** The reason for disconnecting the browser. */
+    reason: string;
+  };
+  /** Emitted when a shared worker is created. */
+  sharedworker: {
+    /** The realm of the shared worker. */
+    realm: SharedWorkerRealm;
+  };
+}> {
+  static async from(session: Session): Promise<Browser> {
+    const browser = new Browser(session);
+    await browser.#initialize();
+    return browser;
+  }
+
+  // keep-sorted start
+  #reason: string | undefined;
+  readonly #userContexts = new Map();
+  readonly session: Session;
+  // keep-sorted end
+
+  private constructor(session: Session) {
+    super();
+
+    // keep-sorted start
+    this.session = session;
+    // keep-sorted end
+
+    this.#userContexts.set('', UserContext.create(this, ''));
+  }
+
+  async #initialize() {
+    // ///////////////////////
+    // Connection listeners //
+    // ///////////////////////
+    const connection = this.#connection;
+    connection.on('script.realmCreated', info => {
+      if (info.type === 'shared-worker') {
+        // TODO: Create a SharedWorkerRealm.
+      }
+    });
+
+    // ///////////////////
+    // Parent listeners //
+    // ///////////////////
+    this.session.once('ended', ({reason}) => {
+      this.#reason = reason;
+      this.emit('disconnected', {reason});
+      this.removeAllListeners();
+    });
+
+    // //////////////////////////////
+    // Asynchronous initialization //
+    // //////////////////////////////
+    const {
+      result: {contexts},
+    } = await connection.send('browsingContext.getTree', {});
```

Comment:
I think this should belong to the UserContext?

Reply:
I think this depends on whether we let `getTree` include a user context ID or not. As of the current draft, this method should get all browsingContexts, independent of the user context, so it should be called on the browser.